src/mod_cache_users.erl

```diff
@@ -0,0 +1,268 @@
+%%% @doc Caches info about non-anonymous users using a queue of ets tables
+-module(mod_cache_users).
+
+-behaviour(gen_mod).
+-behaviour(gen_server).
+
+%% gen_server callbacks
+-export([start_link/2]).
+-export([init/1]).
+-export([handle_call/3]).
+-export([handle_cast/2]).
+-export([handle_info/2]).
+
+%% gen_mod API
+-export([start/2]).
+-export([stop/1]).
+-export([config_spec/0]).
+-export([supported_features/0]).
+
+%% Hooks.
+-export([does_cached_user_exist/3]).
+-export([maybe_put_user_into_cache/3]).
+-export([remove_user/3]).
+-export([remove_domain/3]).
+
+-include("mongoose_config_spec.hrl").
+-include("mongoose.hrl").
+-include("jlib.hrl").
+
+tbl_name() ->
+    mod_cache_users.
+
+-spec tbl_name(mongooseim:host_type()) -> atom().
+tbl_name(HostType) ->
+    gen_mod:get_module_proc(HostType, tbl_name()).
+
+%%====================================================================
+%% API
+%%====================================================================
+
+-spec start(mongooseim:host_type(), gen_mod:module_opts()) -> ok.
+start(HostType, Opts) ->
+    start_server(HostType, Opts),
+    ejabberd_hooks:add(hooks(HostType)),
+    ok.
+
+-spec stop(mongooseim:host_type()) -> ok.
+stop(HostType) ->
+    ejabberd_hooks:delete(hooks(HostType)),
+    stop_server(HostType),
+    ok.
+
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{items = #{
+                       <<"time_to_live">> => #option{type = int_or_infinity,
+                                                     validate = non_negative,
+                                                     format = {kv, ttl}},
+                       <<"number_of_segments">> => #option{type = integer}
+                      }}.
+
+-spec supported_features() -> [atom()].
+supported_features() ->
+    [dynamic_domains].
+
+hooks(HostType) ->
+    [
+     {does_user_exist, HostType, ?MODULE, does_cached_user_exist, 30},
+     {does_user_exist, HostType, ?MODULE, maybe_put_user_into_cache, 70},
+     {remove_user, HostType, ?MODULE, remove_user, 30},
+     {remove_domain, HostType, ?MODULE, remove_domain, 30}
+    ].
+
+%%====================================================================
+%% Hooks
+%%====================================================================
+
+-spec remove_user(Acc :: mongoose_acc:t(),
+                  LUser :: jid:luser(),
+                  LServer :: jid:lserver() | string()) -> mongoose_acc:t().
+remove_user(Acc, LUser, LServer) ->
+    Key = key(LUser, LServer),
+    HostType = mongoose_acc:host_type(Acc),
+    send_to_group(HostType, {remove_user, Key}),
+    Acc.
+
+-spec remove_domain(mongoose_hooks:simple_acc(),
```

Comment:
just a side not, not related to this PR.
we have introduced `remove_domain` but we haven't added `mongoose_hooks:remove_domain/2` execution on `mongoose_domain_api:delete_domain/2` call

Reply:
Ouch, that seems like it should be done, really. Wanna make a ticket for it and put it first in our backlog?